modules/python3gl/src/volumepytoglconverter.cpp

```diff
@@ -0,0 +1,116 @@
+        volumeSrc->getDimensions(), volumeSrc->getDataFormat(), volumeSrc->getSwizzleMask(),
+        volumeSrc->getInterpolation(), volumeSrc->getWrapping(), false);
+
+    const auto tex = volumeDst->getTexture();
+    auto srcSize = static_cast<size_t>(volumeSrc->data().nbytes());
+    auto dstSize = tex->getNumberOfValues() * tex->getSizeInBytes();
+
+    if (srcSize == dstSize &&
+        pybind11::array::c_style == (volumeSrc->data().flags() & pybind11::array::c_style)) {
+        auto src = volumeSrc->data().data(0);
+        volumeDst->getTexture()->initialize(src);
+    } else {
+        throw Exception("Unable to convert from VolumePy to VolumeGL", IVW_CONTEXT);
+    }
+
+    return volumeDst;
+}
+
+void VolumePy2GLConverter::update(std::shared_ptr<const VolumePy> volumeSrc,
+                                  std::shared_ptr<VolumeGL> volumeDst) const {
+    volumeDst->setDimensions(volumeSrc->getDimensions());
+    volumeDst->setSwizzleMask(volumeSrc->getSwizzleMask());
+    volumeDst->setInterpolation(volumeSrc->getInterpolation());
+    volumeDst->setWrapping(volumeSrc->getWrapping());
+
+    const auto tex = volumeDst->getTexture();
+    auto srcSize = static_cast<size_t>(volumeSrc->data().nbytes());
+    auto dstSize = tex->getNumberOfValues() * tex->getSizeInBytes();
+
+    if (srcSize == dstSize &&
+        pybind11::array::c_style == (volumeSrc->data().flags() & pybind11::array::c_style)) {
+        auto src = volumeSrc->data().data(0);
+        volumeDst->getTexture()->upload(src);
+    } else {
+        throw Exception("Unable to convert from VolumePy to VolumeGL", IVW_CONTEXT);
+    }
+}
+
+std::shared_ptr<VolumePy> VolumeGL2PyConverter::createFrom(
+    std::shared_ptr<const VolumeGL> volumeSrc) const {
+    auto volumeDst = std::make_shared<VolumePy>(
+        volumeSrc->getDimensions(), volumeSrc->getDataFormat(), volumeSrc->getSwizzleMask(),
+        volumeSrc->getInterpolation(), volumeSrc->getWrapping());
+
+    const auto tex = volumeSrc->getTexture();
+    auto dstSize = static_cast<size_t>(volumeDst->data().nbytes());
+    auto srcSize = tex->getNumberOfValues() * tex->getSizeInBytes();
+
+    if (srcSize == dstSize &&
+        pybind11::array::c_style == (volumeDst->data().flags() & pybind11::array::c_style)) {
+        auto dst = volumeDst->data().mutable_data(0);
+        volumeSrc->getTexture()->download(dst);
+    } else {
+        throw Exception("Unable to convert from VolumeGL to VolumePy", IVW_CONTEXT);
+    }
+
+    return volumeDst;
+}
+
+void VolumeGL2PyConverter::update(std::shared_ptr<const VolumeGL> volumeSrc,
+                                  std::shared_ptr<VolumePy> volumeDst) const {
+    volumeDst->setDimensions(volumeSrc->getDimensions());
+    volumeDst->setSwizzleMask(volumeSrc->getSwizzleMask());
+    volumeDst->setInterpolation(volumeSrc->getInterpolation());
+    volumeDst->setWrapping(volumeSrc->getWrapping());
+
+    const auto tex = volumeSrc->getTexture();
```

Comment:
is it guaranteed that the correct context is active?

Reply:
No we don't guarantee that